src/devices/bus/qbus/dvk_mx.cpp

```diff
@@ -0,0 +1,745 @@
+
+	DVK "MX" floppy controller (decimal ID 3.057.122)
+
+	Supports four 5.25" drives (single- or double-sided, 40- or 80-track).
+
+	Uses FM encoding and software-defined track format;
+	usually 11 sectors of 256 bytes, with simple additive checksum.
+
+	RT-11 driver name: MX.SYS
+
+	References:
+	https://emuverse.ru/downloads/computers/DVK/docs/KMD/kngmd_MX_2.djvu
+	https://torlus.com/floppy/forum/viewtopic.php?t=1384
+	http://hobot.pdp-11.ru/ukdwk_archive/dwkwebcomplekt/DWKFiles/mx/README.mx1
+
+***************************************************************************/
+
+#include "emu.h"
+#include "dvk_mx.h"
+
+#include "formats/dvk_mx_dsk.h"
+
+#define LOG_WARN    (1U << 1)   // Show warnings
+#define LOG_SHIFT   (1U << 2)   // Shows shift register contents
+#define LOG_REGS    (1U << 3)   // Digital input/output register and data rate select
+#define LOG_STATE   (1U << 4)   // State machine
+#define LOG_LIVE    (1U << 5)   // Live states
+
+#define VERBOSE (LOG_GENERAL | LOG_REGS | LOG_LIVE | LOG_STATE)
+#include "logmacro.h"
+
+#define LOGWARN(...)     LOGMASKED(LOG_WARN, __VA_ARGS__)
+#define LOGSHIFT(...)    LOGMASKED(LOG_SHIFT, __VA_ARGS__)
+#define LOGREGS(...)     LOGMASKED(LOG_REGS, __VA_ARGS__)
+#define LOGLIVE(...)     LOGMASKED(LOG_LIVE, __VA_ARGS__)
+#define LOGSTATE(...)    LOGMASKED(LOG_STATE, __VA_ARGS__)
+
+
+//**************************************************************************
+//  MACROS / CONSTANTS
+//**************************************************************************
+
+#define MXCSR_GETDRIVE(x)  (((x) >> MXCSR_V_DRIVE) & MXCSR_M_DRIVE)
+
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(DVK_MX, dvk_mx_device, "dvk_mx", "DVK MX floppy controller")
+
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  dvk_mx_device - constructor
+//-------------------------------------------------
+
+dvk_mx_device::dvk_mx_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, DVK_MX, tag, owner, clock)
+	, device_qbus_card_interface(mconfig, *this)
+{
+	memset(&cur_live, 0x00, sizeof(cur_live));
+	cur_live.tm = attotime::never;
+	cur_live.state = IDLE;
+	cur_live.next_state = -1;
+}
+
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void dvk_mx_device::device_start()
+{
+	for (int i = 0; i != 4; i++)
+	{
+		char name[2];
+		flopi[i].tm = timer_alloc(FUNC(dvk_mx_device::update_floppy), this);
+		flopi[i].id = i;
+		name[0] = '0' + i;
+		name[1] = 0;
+		floppy_connector *con = subdevice<floppy_connector>(name);
```

Comment:
Please go for a required_device_array, it's rather nicer.

Reply:
Not sure how to use it here, tbh.

Reply:
You have a member like:
```c++
	required_device_array<floppy_connector, 4> m_connectors;
```

Initialise it like:
```
	m_connectors(*this, "%u", 0U)
```

And then you can just do like `m_connectors[i]->whatever()` and use it in place of literal tag string in the machine configuration function as well.